plugins/de.cau.cs.kieler.klighd.lsp/src/de/cau/cs/kieler/klighd/lsp/KGraphDiagramServer.xtend

```diff
@@ -0,0 +1,599 @@
+            val SModelCloner cloner = getSModelCloner();
+            val SModelRoot newRoot = cloner.clone(model);
+            synchronized(modelLock) {
+                newRoot.setRevision(revision + 1);
+                revision++
+                currentRoot = newRoot;
+            }
+            // the actual layout is performed in doSubmitModel
+            doSubmitModel(newRoot, true, action);
+        }
+    }
+    
+    /**
+     * Called when a {@link PerformActionAction} is received.
+     * Invokes the contained KlighD {@link IAction} on the current {@link KNode KGraph}.
+     * May cause a {@link UpdateModelAction} to be sent back to the client with an updated model.
+     */
+    protected def handle(PerformActionAction action) {
+        synchronized (diagramState) {
+            
+            val sourceUrl = diagramState.getURIString(clientId)
+            val k2sMap = diagramState.getKGraphToSModelElementMap(sourceUrl)
+            val kGraphElement = KGraphElementIDGenerator.findElementById(k2sMap, action.KGraphElementId)
+            val kRendering = KRenderingIDGenerator.findRenderingById(kGraphElement, action.KRenderingId)
+            
+            val klighdAction = KlighdDataManager.instance.getActionById(action.actionId)
+            val viewer = diagramState.viewer
+            val actionContext = new ActionContext(viewer, null, kGraphElement, kRendering)
+            val actionResult = klighdAction.execute(actionContext)
+            if (actionResult.needsSynthesis) {
+                updateDiagram()
+            } else if (actionResult.actionPerformed) {
+                updateLayout()
+            }
+        }
+    }
+    
+    /**
+     * Called when a {@link SetSynthesisAction} is received.
+     * Configures this diagram server to use the synthesis given in the message in future runs to generate a diagram
+     * and invokes an update.
+     */
+    protected def handle(SetSynthesisAction action) {
+        synchronized (diagramState) {
+            val sourceUrl = diagramState.getURIString(clientId)
+            diagramState.putSynthesisId(sourceUrl, action.id)
+            this.newModel = true
+            updateDiagram()
+        }
+    }
+    
+    /**
+     * Called when a {@link CheckedImagesAction} is received.
+     * Tells the server that the images on the client have been checked if they are cached and requests any non-cached
+     * images to be sent.
+     * The requested images are then sent to the client.
+     */
+    protected def handle(CheckedImagesAction action) {
+        synchronized (modelLock) {
+            if (action.notCached.empty) {
+                imagesUpdated = true
+            } else {
+                val images = new ArrayList<Pair<String, String>>
+                for (notCached : action.notCached) {
+                    try {
+                        val bundleAndPath = notCached.split(":", 2)
```

Comment:
No, this should be two fields in the action containing the bundle name and the path. There however is a bigger problem here: This will only work with the platform running. The Eclipseless version of the LS always throws an exception here when resolving the bundle. Needs a more general solution.